### PR TITLE
Studio: Add Song of the Day functionality

### PR DESCRIPTION
**Pull request by panpawn**
Requested by #Don't Lose

### CODE REVIEW

File: chat-plugins/thestudio.js [hunk 1]
```diff
@@ -257,11 +257,35 @@ let commands = {
...
+		if (!this.canTalk()) return;
+
+		if (target === 'off' || target === 'disable') {
+			room.chatRoomData.songOfTheDay = false;
+			Rooms.global.writeChatRoomData();
+			return this.privateModCommand(`(${user.name} has disabled the Song of the Day.)`);
+		}
+
+		let title = target.split(',')[0].trim();
+		let link = target.split(',')[1].trim();
```
Comment (Slayer95) [inline chat-plugins/thestudio.js]: This crashes if the target has no comma

File: chat-plugins/thestudio.js [hunk 2]
```diff
@@ -257,11 +257,35 @@ let commands = {
...
+
+		room.chatRoomData.songOfTheDay = `<a href="${link}" title="${title}">${Chat.escapeHTML(title)}</a>.`;
```
Comment (Slayer95) [inline chat-plugins/thestudio.js]: The attributes need HTML escaping

File: chat-plugins/thestudio.js [hunk 3]
```diff
@@ -257,11 +257,35 @@ let commands = {
...
+		Rooms.global.writeChatRoomData();
+		this.privateModCommand(`(${user.name} set the Song of the Day to ${title} (Link: ${link}.)`);
+	},
+	songhelp: [
+		"/aotd song [song], [link] - Sets the Song of the Day. Requires % @ * # & ~",
+		"/oatd song [off|disable] - Turns the Song of the Day off. Requires 5 @ * # & ~",
```
Comment (Slayer95) [inline chat-plugins/thestudio.js]: Tipo

> Requires 5 @ * # &

File: chat-plugins/thestudio.js [hunk 4]
```diff
@@ -285,5 +309,7 @@ exports.commands = {
 		"- /aotd set [artist] - Set the Artist of the Day. Requires: % @ * # & ~",
 		"- /aotd quote - View the current Artist Quote of the Day.",
 		"- /aotd quote [quote] - Set the Artist Quote of the Day. Requires: # & ~",
+		"- /aotd song [song], [link] - Sets the Song of the Day. Requires % @ * # & ~",
+		"- /oatd song [off|disable] - Turns the Song of the Day off. Requires 5 @ * # & ~",
```
Comment (Slayer95) [inline chat-plugins/thestudio.js]: Typo in the namespace

Comment (Slayer95) [inline chat-plugins/thestudio.js]: Also in the permissions

**Comment by panpawn** [comment]
It's good to have you back @Slayer95 :o)

I believe I've updated this with all of the requested changes thus far, I must of been really tired when I made this originally...

File: chat-plugins/thestudio.js [hunk 5]
```diff
@@ -257,11 +257,37 @@ let commands = {
...
+
+		if (target === 'off' || target === 'disable') {
+			room.chatRoomData.songOfTheDay = false;
+			Rooms.global.writeChatRoomData();
+			return this.privateModCommand(`(${user.name} has disabled the Song of the Day.)`);
+		} else if (target.includes(',')) {
+			let title = target.split(',')[0].trim();
+			let link = target.split(',')[1].trim();
+
+			room.chatRoomData.songOfTheDay = `<a href="${link}" title="${Chat.escapeHTML(title)}">${Chat.escapeHTML(title)}</a>.`;
```
Comment (Slayer95) [inline chat-plugins/thestudio.js]: The ``href`` *still* needs escaping...

Comment (Slayer95) [inline chat-plugins/thestudio.js]: ``Chat.html`` should do the job here, actually.

File: chat-plugins/thestudio.js [hunk 6]
```diff
@@ -257,11 +257,37 @@ let commands = {
...
+		"/aotd song [song], [link] - Sets the Song of the Day. Requires % @ * # & ~",
+		"/aotd song [off|disable] - Turns the Song of the Day off. Requires % @ * # & ~",
+	],
 
 	'': function (target, room) {
 		if (room.id !== 'thestudio') return this.errorReply('This command can only be used in The Studio.');
 		if (!room.chatRoomData || !this.runBroadcast()) return false;
-		this.sendReplyBox("The Artist of the Day " + (room.chatRoomData.artistOfTheDay ? "is " + room.chatRoomData.artistOfTheDay + "." : "has not been set yet."));
+		let song = (room.chatRoomData.songOfTheDay ? '<br />The Song of the Day is ' + room.chatRoomData.songOfTheDay : '');
+		this.sendReplyBox("The Artist of the Day " + (room.chatRoomData.artistOfTheDay ? "is " + room.chatRoomData.artistOfTheDay + "." : "has not been set yet.") + song);
```
Comment (Slayer95) [inline chat-plugins/thestudio.js]: While you are at this, it seems that ``artistOfTheDay`` may contain special chars, but I may be wrong

**Comment by panpawn** [comment]
Updated @Slayer95 

You were right, there was a lack of escaping with the AotD.

I also updated this to use ``Chat#html`` for where I use that template string.

File: chat-plugins/thestudio.js (hunk 6 above)
Comment (panpawn) [inline chat-plugins/thestudio.js]: No, you're right! Fixed :p

**Comment by Slayer95** [comment]
I'm not a fan of storing escaped strings / markup, but I'll let someone else champion that.

**Comment by Zarel** [comment]
I agree, escape on display, not on storage.

**Comment by panpawn** [comment]
Done

File: chat-plugins/thestudio.js [hunk 7]
```diff
@@ -257,11 +257,37 @@ let commands = {
...
+
+		if (target === 'off' || target === 'disable') {
+			room.chatRoomData.songOfTheDay = false;
+			Rooms.global.writeChatRoomData();
+			return this.privateModCommand(`(${user.name} has disabled the Song of the Day.)`);
+		} else if (target.includes(',')) {
+			let title = target.split(',')[0].trim();
+			let link = target.split(',')[1].trim();
+
+			room.chatRoomData.songOfTheDay = Chat.html`<a href="${link}" title="${title}">${title}</a>.`;
```
Comment (Zarel) [inline chat-plugins/thestudio.js]: This should probably be called `songOfTheDayHTML` if it contains HTML, as it does.

**Comment by panpawn** [comment]
done

File: chat-plugins/thestudio.js [hunk 8]
```diff
@@ -257,11 +257,37 @@ let commands = {
...
+		"/aotd song [song], [link] - Sets the Song of the Day. Requires % @ * # & ~",
+		"/aotd song [off|disable] - Turns the Song of the Day off. Requires % @ * # & ~",
+	],
 
 	'': function (target, room) {
 		if (room.id !== 'thestudio') return this.errorReply('This command can only be used in The Studio.');
 		if (!room.chatRoomData || !this.runBroadcast()) return false;
-		this.sendReplyBox("The Artist of the Day " + (room.chatRoomData.artistOfTheDay ? "is " + room.chatRoomData.artistOfTheDay + "." : "has not been set yet."));
+		let song = (room.chatRoomData.songOfTheDayHTML ? '<br />The Song of the Day is ' + room.chatRoomData.songOfTheDayHTML : '');
+		this.sendReplyBox("The Artist of the Day " + (room.chatRoomData.artistOfTheDay ? "is " + room.chatRoomData.artistOfTheDay + "." : "has not been set yet.") + song);
```
Comment (Zarel) [inline chat-plugins/thestudio.js]: This does not appear to correctly escape the artist of the day.

Comment (panpawn) [inline chat-plugins/thestudio.js]: ops

**Comment by panpawn** [comment]
Updated

**Pull request merged**